http/src/v0/root_files/add.rs

```diff
@@ -15,97 +19,213 @@ pub(super) async fn add_inner<T: IpfsTypes>(
 
-        if field_name != "file" {
-            return Err(StringError::from(format!("unsupported field: {}", field_name)).into());
-        }
+    let st = add_stream(ipfs, stream);
 
-        let filename = field
-            .filename()
-            .map_err(|e| StringError::from(format!("unparseable filename: {}", e)))?
-            .to_string();
+    // TODO: we could map the errors into json objects at least? (as we cannot return them as
+    // trailers)
 
-        let mut adder = FileAdder::default();
-        let mut total = 0u64;
+    let body = crate::v0::support::StreamResponse(st);
+
+    Ok(body)
+}
 
-        loop {
-            let next = field
-                .try_next()
-                .await
-                .map_err(|e| StringError::from(format!("IO error: {}", e)))?;
+#[derive(Debug)]
+enum AddError {
+    Parsing(MultipartError),
+    Header(MultipartError),
+    InvalidFilename(std::str::Utf8Error),
+    UnsupportedField(String),
+    UnsupportedContentType(String),
+    ResponseSerialization(serde_json::Error),
+    Persisting(ipfs::Error),
+    TreeGathering(TreeBuildingFailed),
+    TreeBuilding(TreeConstructionFailed),
+}
 
-            match next {
-                Some(next) => {
-                    let mut read = 0usize;
-                    while read < next.len() {
-                        let (iter, used) = adder.push(&next.slice(read..));
-                        read += used;
+impl From<MultipartError> for AddError {
+    fn from(e: MultipartError) -> AddError {
+        AddError::Parsing(e)
+    }
+}
 
-                        let maybe_tuple = import_all(&ipfs, iter).await.map_err(|e| {
-                            StringError::from(format!("Failed to save blocks: {}", e))
-                        })?;
+impl fmt::Display for AddError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // TODO
+        write!(fmt, "{:?}", self)
+    }
+}
 
-                        total += maybe_tuple.map(|t| t.1).unwrap_or(0);
+impl std::error::Error for AddError {}
+
+fn add_stream<St, E>(
+    ipfs: Ipfs<impl IpfsTypes>,
+    mut fields: MultipartStream<St, E>,
+) -> impl Stream<Item = Result<Bytes, AddError>> + Send + 'static
+where
+    St: Stream<Item = Result<Bytes, E>> + Send + Unpin + 'static,
+    E: Into<anyhow::Error> + Send + 'static,
+{
+    async_stream::try_stream! {
+        // TODO: wrap-in-directory option
+        let mut tree = BufferingTreeBuilder::default();
+        let mut buffer = BytesMut::new();
+
+        while let Some(mut field) = fields
+            .try_next()
+            .await?
+        {
+
+            let field_name = field.name().map_err(AddError::Header)?;
+            let filename = field.filename().map_err(AddError::Header)?;
+            let filename = percent_encoding::percent_decode_str(filename)
+                .decode_utf8()
+                .map(|cow| cow.into_owned())
+                .map_err(AddError::InvalidFilename)?;
+
```

Comment:
missing TODO: parse metadata

Reply:
though probably good to leave it for another PR, there's already a bunch of stuff here :)

Reply:
Leaving this out of this PR, shouldn't be too hard to implement.